homeassistant/components/tahoma/cover.py

```diff
@@ -0,0 +1,359 @@
+            CORE_SLATS_OPEN_CLOSED_STATE,
+            CORE_OPEN_CLOSED_PARTIAL_STATE,
+            CORE_OPEN_CLOSED_PEDESTRIAN_STATE,
+            CORE_OPEN_CLOSED_UNKNOWN_STATE,
+            MYFOX_SHUTTER_STATUS_STATE,
+        )
+
+        if state is not None:
+            return state == STATE_CLOSED
+
+        if self.current_cover_position is not None:
+            return self.current_cover_position == 0
+
+        if self.current_cover_tilt_position is not None:
+            return self.current_cover_tilt_position == 0
+
+        return None
+
+    @property
+    def device_class(self):
+        """Return the class of the device."""
+        return (
+            TAHOMA_COVER_DEVICE_CLASSES.get(self.device.widget)
+            or TAHOMA_COVER_DEVICE_CLASSES.get(self.device.ui_class)
+            or DEVICE_CLASS_BLIND
+        )
+
+    @property
+    def icon(self):
+        """Return the icon to use in the frontend, if any."""
+        if (
+            self.executor.has_state(CORE_PRIORITY_LOCK_TIMER_STATE)
+            and self.executor.select_state(CORE_PRIORITY_LOCK_TIMER_STATE) > 0
+        ):
+            if self.executor.select_state(IO_PRIORITY_LOCK_ORIGINATOR_STATE) == "wind":
+                return ICON_WEATHER_WINDY
+            return ICON_LOCK_ALERT
+
+        return None
+
+    async def async_open_cover(self, **_):
+        """Open the cover."""
+        await self.executor.async_execute_command(
+            self.executor.select_command(*COMMANDS_OPEN)
+        )
+
+    async def async_open_cover_tilt(self, **_):
+        """Open the cover tilt."""
+        await self.executor.async_execute_command(
+            self.executor.select_command(*COMMANDS_OPEN_TILT)
+        )
+
+    async def async_close_cover(self, **_):
+        """Close the cover."""
+        await self.executor.async_execute_command(
+            self.executor.select_command(*COMMANDS_CLOSE)
+        )
+
+    async def async_close_cover_tilt(self, **_):
+        """Close the cover tilt."""
+        await self.executor.async_execute_command(
+            self.executor.select_command(*COMMANDS_CLOSE_TILT)
+        )
+
+    async def async_stop_cover(self, **_):
+        """Stop the cover."""
+        await self.async_cancel_or_stop_cover(
+            COMMANDS_OPEN + COMMANDS_SET_POSITION + COMMANDS_CLOSE,
+            COMMANDS_STOP,
+        )
+
+    async def async_stop_cover_tilt(self, **_):
+        """Stop the cover tilt."""
+        await self.async_cancel_or_stop_cover(
+            COMMANDS_OPEN_TILT + COMMANDS_SET_TILT_POSITION + COMMANDS_CLOSE_TILT,
+            COMMANDS_STOP_TILT,
+        )
+
+    async def async_cancel_or_stop_cover(self, cancel_commands, stop_commands) -> None:
+        """Cancel running execution or send stop command."""
+        # Cancelling a running execution will stop the cover movement
+        # Retrieve executions initiated via Home Assistant from Data Update Coordinator queue
+        exec_id = next(
```

Comment:
Looks like we do this more than once...Can this be its own function?

Reply:
This specific functionality is only used once. We already moved this logic from `stop_cover*` to `async_cancel_or_stop_cover`, so not sure if further decoupling is possible and would make sense.